build-support/native-image/generate-native-image-for-pants-targets.bash

```diff
@@ -0,0 +1,205 @@
+      cp -v {,lib}native-image-agent.dylib || return "$?"
+    else
+      cp -v {,lib}native-image-agent.so || return "$?"
+    fi >&2 || return "$?"
+  fi
+  echo "$desired_agent_dylib_path"
+  popd >&2
+}
+
+# run with tracing on
+
+function run_zinc_compile_with_tracing {
+  agent_lib_path="$(build_agent_dylib)"
+  if [[ ! -f "$GENERATED_CONFIG_JSON_FILE" ]]; then
+    rm -rfv "$GENERATED_CONFIG_DIR"
+    mkdir -v "$GENERATED_CONFIG_DIR"
+    # TODO: jvm options are ignored from the command line, this is a bug and should be fixed.
+    export PANTS_COMPILE_ZINC_JVM_OPTIONS="+['-agentlib:native-image-agent=config-merge-dir=${GENERATED_CONFIG_DIR}']"
+    # NB: --worker-count=1 is because the `config-merge-dir` option to the native-image agent will
+    # clobber symbols from concurrent compiles.
+    MODE=debug ./pants -ldebug \
+            --no-zinc-native-image \
+            --resolve-coursier-capture-snapshots \
+            compile.zinc \
+            --execution-strategy=hermetic \
+            --worker-count=1 \
+            --no-incremental \
+            --no-use-classpath-jars \
+            --native-image-agent-library-dir="$(dirname "$agent_lib_path")" \
+            --cache-ignore \
+            "$@"
+    unset PANTS_COMPILE_ZINC_JVM_OPTIONS
+  fi >&2
+  normalize_path "$GENERATED_CONFIG_DIR"
+}
+
+# analyze tracing -- trim reflection down to just macros, then find those macros' targets with
+# classmap
+
+function trim_reflection_config_to_just_macros {
+  # Scala macros are the only entries in reflect-config.json that appear to matter for the purposes
+  # of compilation, so we filter down to only those.
+  # TODO: do compiler plugins look different?
+  jq '[.[] | select(.fields[]? | .name == "MODULE$")]' \
+     <"$GENERATED_CONFIG_JSON_FILE" \
+     >.tmp-reflect-config \
+    && mv -v \
+          .tmp-reflect-config \
+          "$GENERATED_CONFIG_JSON_FILE"
+}
+
+function extract_macro_target_names_from_pants {
+  ./pants classmap "$@" >.tmp-classmap || return "$?"
+
+  # Get the class names of all the macro usages, then filter down the above `./pants classmap` to
+  # find the targets which provide those classes.
+  jq -r '.[] | .name' <"$GENERATED_CONFIG_JSON_FILE" | while read -r class_name; do
+    grep -F "$class_name" .tmp-classmap \
+      | sed -Ee 's#^.* (.*)$#\1#g'
+  done | sort -u
+}
+
+# create a BUILD file of the appropriate dependencies
+
+function template_BUILD_file {
+  cat <<EOF
+# This file was generated by the scripts in https://github.com/pantsbuild/pants/tree/master/build-support/native-image/.
+# DO NOT EDIT!!!
+
+jvm_binary(
+  name='generated-native-image-binary',
+  main='org.pantsbuild.zinc.compiler.Main',
+  dependencies=[
+    $(sed -Ee 's#^(.*)$#"\1",#g')
+  ],
+)
+
+jvm_app(
+  name='native-image-deps',
+  basename='gen-native-image-basename',
+  binary=':generated-native-image-binary',
+  deployjar=True,
+)
+EOF
+}
+
+# generate the deps jar and reflect config at known locations!!!
+
```

Comment:
It feels like this script needs a bit more baking. Is it worth extracting this file out of this PR, finishing the design for how the native image will be maintained, and then revisiting it?

Reply:
I think that this script should absolutely be considered a prototype given:
(1) it will be significantly less complex once we can upstream our native-image fixes into oracle/graal#1448 via #7955 
(2) it's difficult to write reproducible bash scripts, so it would probably need breaking down into component rules anyway

And (3) I think "here is a script that works, with clear instructions on how to run it from any repo" feels like a great way to decouple "being able to generate native-images" from "here is a working implementation in pants".

Reply:
I'm not sure that answers my question... do you think it would be reasonable to break this out and finish the UX design before landing this particular file? I'm just talking about deferring `generate-native-image-for-pants-targets.bash`, not about `build-zinc-native-image.bash`: the latter should land here I think.